optimade/models/structures.py

```diff
@@ -22,7 +21,7 @@
 )
 from optimade.server.warnings import MissingExpectedField
 
-EXTENDED_CHEMICAL_SYMBOLS = CHEMICAL_SYMBOLS + EXTRA_SYMBOLS
+EXTENDED_CHEMICAL_SYMBOLS = set(CHEMICAL_SYMBOLS + EXTRA_SYMBOLS)
```

Comment:
I don't see how making this a set does anything?

Reply:
It is only ever used here:

https://github.com/Materials-Consortia/optimade-python-tools/blob/c6f76d8ec238419642619681ff6e7fbb87a42d65/optimade/models/structures.py#L148-L151

so every chemical symbol is going to do a O(N) look-up within the same list, may as well pre-compute the set once to get to O(1)

Reply:
I did the same for `CHEMICAL_SYMBOLS` but it changes all of the error messages in our tests, so we need to spend a bit more time on that

Reply:
> Hmm. I don't think I completely follow this still (since I don't know the underlying Python workflow for checking part-hood and whether there's a difference between checking part-hood of lists of sets),

It's just a set, not a list of sets.

> On another note we should use the notation `if v not in EXTENDED_CHEMICAL_SYMBOLS`, right? But that's not so important.

Yeah, I'm not sure why this line is like that, so let's change it now too.

Reply:
Ah - so are `CHEMICAL_SYMBOLS` and `EXTRA_SYMBOLS` already sets then? Yeah okay - but then why not make a union?

Reply:
They are both lists, so adding them together and casting to a set is equivalent to making two individual sets and taking a union.

```python
>>> from optimade.models.utils import CHEMICAL_SYMBOLS, EXTRA_SYMBOLS
>>> set(CHEMICAL_SYMBOLS + EXTRA_SYMBOLS)
{'Ds', 'Sg', 'Cd', 'U', 'Be', 'Tm', 'Mc', 'Hs', 'Fe', 'Pt', 'As', 'Eu', 'Fl', 'Lu', 'Ge', 'Zr', 'Zn', 'Y', 'F', 'Pa', 'Nh', 'Np', 'Gd', 'Sr', 'Th', 'Kr', 'Pr', 'Ca', 'Dy', 'Ir', 'Rb', 'Fm', 'Am', 'Re', 'Co', 'Nb', 'Er', 'Sc', 'Os', 'Hg', 'N', 'Pu', 'Mg', 'Ne', 'K', 'Po', 'W', 'Tc', 'Es', 'Au', 'Ti', 'Pb', 'B', 'Na', 'Bk', 'C', 'Tl', 'Ce', 'Pm', 'Li', 'Ac', 'Lv', 'Nd', 'Br', 'No', 'Ta', 'Og', 'Sn', 'At', 'Ba', 'Ho', 'Cl', 'La', 'P', 'Ga', 'Ni', 'Db', 'Cm', 'Rf', 'Fr', 'Sb', 'S', 'Si', 'Ar', 'Lr', 'Mt', 'O', 'V', 'Yb', 'Cf', 'Rh', 'Te', 'In', 'Tb', 'Rn', 'Cs', 'Ru', 'Mn', 'Ra', 'Pd', 'Bh', 'Ts', 'He', 'vacancy', 'Md', 'Xe', 'Rg', 'X', 'H', 'Mo', 'Se', 'Cn', 'Cu', 'Sm', 'I', 'Al', 'Hf', 'Cr', 'Bi', 'Ag'}
```

Reply:
But that's what I don't understand, why then cast to a set, a list would do just fine, right? Or is there some advantage to using set in this case that I'm unaware of?

Reply:
Once you have constructed the set, look-ups are basically free. From some rudimentary benchmarking, i.e. looping over a list of 600 species and checking if they are in the allowed values, pre-computing a set (which takes a few µs on first import) and doing look-ups against it is about 40x faster than the same using the list:

```python
In [26]: %timeit find_symbols(bigger_list, symbol_set)
23.1 µs ± 887 ns per loop (mean ± std. dev. of 7 runs, 10000 loops each)

In [27]: %timeit find_symbols(bigger_list, symbol_list)
889 µs ± 20 µs per loop (mean ± std. dev. of 7 runs, 1000 loops each)
```

where

```python
def find_symbols(test, allowed):
    for symbol in test:
        if symbol not in allowed:
            pass
```

and `bigger_list = 100*["H", "C", "N", "X", "vacancy", "D"]`, `symbol_list = CHEMICAL_SYMBOLS + EXTRA_SYMBOLS` and `symbol_set = set(symbol_list)`. Speed-up is more pronounced if you use random chemical symbols.

Reply:
Ah!! Didn't know this 😮 amazing! And makes sense since a set is guaranteed to have unique individuals and one doesn't care about order. Excellent!